### PR TITLE
Sum factorisation and delta elimination for coefficients

### PR DESCRIPTION
**Pull request by miklos1**
Ticks two boxes in #75. Can't merge currently without disabling COFFEE because of coneoproject/COFFEE#100.

### CODE REVIEW

File: gem/optimise.py [hunk 1]
```diff
@@ -191,6 +196,168 @@ def select_expression(expressions, index):
...
+            if flops < best_flops:
+                expression = expr
+                best_flops = flops
+    else:
+        # Cheap heuristic
+        def key(factor):
+            return len(set(sum_indices) & set(factor.free_indices))
+        ordering = sorted(factors, key=key)
+
+        # FIXME: Log this unexpected case.
```
Comment (miklos1) [inline gem/optimise.py]: GEM doesn't have a logger, and I don't want to introduce reverse dependency on TSFC.

Comment (wence-) [inline gem/optimise.py]: Could you fix this by requiring the caller of gem to provide a logger?

File: gem/gem.py [hunk 2]
```diff
@@ -553,26 +553,27 @@ def __new__(cls, expression, multiindex):
...
         # Sum zeros
         assert not summand.shape
         if isinstance(summand, Zero):
             return summand
 
-        # Sum a single expression
-        if index.extent == 1:
-            return Indexed(ComponentTensor(summand, (index,)), (0,))
+        # No indices case
+        multiindex = tuple(multiindex)
```
Comment (miklos1) [inline gem/gem.py]: The ordering of these indices logically shouldn't matter.

File: gem/optimise.py [hunk 3]
```diff
@@ -191,6 +196,175 @@ def select_expression(expressions, index):
...
+    # Flatten a product tree
+    sum_indices = []
+    factors = []
+
+    queue = deque([expression])
+    while queue:
+        expr = queue.popleft()
+        if isinstance(expr, IndexSum):
+            queue.append(expr.children[0])
+            sum_indices.extend(expr.multiindex)
```
Comment (miklos1) [inline gem/optimise.py]: However, presumably here, we rely on the indices coming out in the same order on all nodes for deterministic code generation.

**Comment by miklos1** [comment]
This does not work well with derivatives.

**Pull request closed without merging**